templates/nft-template/move/sources/nft_launchpad.move

```diff
@@ -0,0 +1,466 @@
+    use std::option;
+    use std::signer;
+    use std::string;
+    use std::vector;
+    use aptos_std::simple_map;
+    use aptos_std::string_utils;
+    use aptos_framework::aptos_account;
+    use aptos_framework::event;
+    use aptos_framework::object;
+    use aptos_token_objects::collection;
+    use aptos_token_objects::royalty;
+    use aptos_token_objects::token;
+
+    use minter::token_components;
+    use minter::mint_stage;
+    use minter::collection_components;
+
+    /// Sender is not admin
+    const E_NOT_ADMIN: u64 = 1;
+    /// Sender is not pending admin
+    const E_NOT_PENDING_ADMIN: u64 = 2;
+    /// Only admin can update mint fee collector
+    const E_ONLY_ADMIN_CAN_UPDATE_MINT_FEE_COLLECTOR: u64 = 3;
+    /// Only admin can create collection
+    const E_ONLY_ADMIN_CAN_CREATE_COLLECTION: u64 = 4;
+    /// No mint limit
+    const E_NO_MINT_LIMIT: u64 = 5;
+    /// Mint limit reached
+    const E_MINT_LIMIT_REACHED: u64 = 6;
+    /// No active mint stages
+    const E_NO_ACTIVE_STAGES: u64 = 7;
+
+    const ALLOWLIST_MINT_STAGE_CATEGORY: vector<u8> = b"Allowlist mint stage";
+    const PUBLIC_MINT_MINT_STAGE_CATEGORY: vector<u8> = b"Public mint mint stage";
+
+    #[event]
+    struct CreateCollectionEvent has store, drop {
+        creator_addr: address,
+        collection_owner_obj: object::Object<CollectionOwnerObjConfig>,
+        collection_obj: object::Object<collection::Collection>,
+        max_supply: option::Option<u64>,
+        name: string::String,
+        uri: string::String,
+        pre_mint_amount: u64,
+        allowlist: option::Option<vector<address>>,
+        allowlist_start_time: option::Option<u64>,
+        allowlist_end_time: option::Option<u64>,
+        allowlist_mint_limit_per_addr: option::Option<u64>,
+        allowlist_mint_fee_per_nft: option::Option<u64>,
+        public_mint_start_time: option::Option<u64>,
+        public_mint_end_time: option::Option<u64>,
+        public_mint_limit_per_addr: option::Option<u64>,
+        public_mint_fee_per_nft: option::Option<u64>,
+    }
+
+    #[event]
+    struct MintNftEvent has store, drop {
+        collection_obj: object::Object<collection::Collection>,
+        nft_obj: object::Object<token::Token>,
+        recipient_addr: address,
+        mint_fee: u64,
+    }
+
+    /// Unique per collection
+    /// We need this object to own the collection object instead of contract directly owns the collection object
+    /// This helps us avoid address collision when we create multiple collections with same name
+    struct CollectionOwnerObjConfig has key {
+        /// Only thing it stores is the link to collection object
+        collection_obj: object::Object<collection::Collection>,
+        extend_ref: object::ExtendRef,
+    }
+
+    /// Unique per collection
+    struct CollectionConfig has key {
+        /// Key is stage, value is mint fee denomination
+        mint_fee_per_nft_by_stages: simple_map::SimpleMap<string::String, u64>,
```

Comment:
We should be using the coin_payment module here: 
https://github.com/aptos-labs/token-minter/blob/main/token-minter/sources/modules/coin_payment.move

Reply:
I used that initially, but think it's an overkill here